### PR TITLE
Layoutlm onnx support (Issue #13300)

### PR DESCRIPTION
**Pull request by nishprabhu**
# What does this PR do?

This PR extends ONNX support to LayoutLM as explained in https://huggingface.co/transformers/serialization.html?highlight=onnx#converting-an-onnx-model-using-the-transformers-onnx-package

Fixes Issue #13300

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [X] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [X] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [X] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@NielsRogge @mfuntowicz @LysandreJik

### CODE REVIEW

File: src/transformers/onnx/features.py [hunk 1]
```diff
@@ -78,6 +79,7 @@ class FeaturesManager:
             "sequence-classification-with-past",
             onnx_config_cls=GPTNeoOnnxConfig,
         ),
+        "layoutlm": supported_features_mapping("default", onnx_config_cls=LayoutLMOnnxConfig),
```
Comment (michaelbenayoun) [inline src/transformers/onnx/features.py:82]: Is there any limitation that makes support for masked language model, sequence classification and token classification tasks complicated?

If that's the case, then we can go forward and add support for these tasks later, but if not you can add those now:

```suggestion
        "layoutlm": supported_features_mapping(
            "default", 
            "masked-lm",
            "sequence-classification",          
            "token-classification",                               
            onnx_config_cls=LayoutLMOnnxConfig,                  
        )   
```
This also means adding a key "masked-lm" mapping to the AutoModelForMaskedLM class to the _TASK_TO_AUTOMODELS dict.

File: src/transformers/models/layoutlm/configuration_layoutlm.py [hunk 2]
```diff
@@ -125,3 +129,92 @@ def __init__(
...
+        dummy_input = [" ".join([tokenizer.unk_token]) * seq_length] * batch_size
+        input_dict = dict(tokenizer(dummy_input, return_tensors=framework))
+
+        # Generate a dummy bbox
+        box = [48, 84, 73, 128]
+
+        if not framework == TensorType.PYTORCH:
+            raise NotImplementedError("Exporting LayoutLM to ONNX is currently only supported for PyTorch.")
+
+        import torch
```
Comment (michaelbenayoun) [inline src/transformers/models/layoutlm/configuration_layoutlm.py:211]: You should make sure torch is available:
```suggestion
        if not is_torch_available():
              raise ValueError("Cannot generate dummy inputs without PyTorch installed.")
          else:
              import torch
```

File: src/transformers/models/layoutlm/configuration_layoutlm.py [hunk 3]
```diff
@@ -125,3 +129,92 @@ def __init__(
...
+                ("input_ids", {0: "batch", 1: "sequence"}),
+                ("bbox", {0: "batch", 1: "sequence"}),
+                ("attention_mask", {0: "batch", 1: "sequence"}),
+                ("token_type_ids", {0: "batch", 1: "sequence"}),
+            ]
+        )
+
+    @property
+    def outputs(self) -> Mapping[str, Mapping[int, str]]:
+        return OrderedDict(
```
Comment (michaelbenayoun) [inline src/transformers/models/layoutlm/configuration_layoutlm.py:157]: If support for other tasks is possible (check comment below), you should use something like:
```python
common_outputs = super().outputs
```
That way, the output names will be conditioned on the task.

File: src/transformers/onnx/features.py (hunk 1 above)
Comment (nishprabhu) [reply src/transformers/onnx/features.py:82]: I don't see any limitation that would complicate those tasks. I think we can add them now. @NielsRogge what do you think?

Comment (NielsRogge) [reply src/transformers/onnx/features.py:82]: Yes you can add them.

File: src/transformers/models/layoutlm/configuration_layoutlm.py (hunk 3 above)
Comment (nishprabhu) [reply src/transformers/models/layoutlm/configuration_layoutlm.py]: Thanks for pointing this out. I have removed this property. @michaelbenayoun

File: src/transformers/models/layoutlm/configuration_layoutlm.py (hunk 2 above)
Comment (nishprabhu) [reply src/transformers/models/layoutlm/configuration_layoutlm.py:188]: Thanks! I've added that check. @michaelbenayoun

File: src/transformers/onnx/__init__.py [hunk 4]
```diff
@@ -15,4 +15,4 @@
 
 from .config import EXTERNAL_DATA_FORMAT_SIZE_LIMIT, OnnxConfig, OnnxConfigWithPast, PatchingSpec
 from .convert import export, validate_model_outputs
-from .utils import ParameterFormat, compute_serialized_parameters_size
+from .utils import ParameterFormat, compute_effective_axis_dimension, compute_serialized_parameters_size
```
Comment (michaelbenayoun) [inline src/transformers/onnx/__init__.py:18]: Is this import still needed as if I understand correctly it was previously used by the `generate_dummy_inputs` method?

**Comment by michaelbenayoun** [comment]
Looks almost ready, great work @nishprabhu!!
Thank you for your contribution.

File: src/transformers/onnx/__init__.py (hunk 4 above)
Comment (nishprabhu) [reply src/transformers/onnx/__init__.py]: Yes, I forgot about that. I have removed it now. Thanks!

**Review by michaelbenayoun — APPROVED** [review]

**Review by LysandreJik — APPROVED** [review]
This looks good to me. Do you want to have a look @NielsRogge @mfuntowicz ?

**Comment by nishprabhu** [comment]
Hi, any update about this? Any other changes required?
@LysandreJik @mfuntowicz @NielsRogge

File: src/transformers/models/layoutlm/configuration_layoutlm.py [hunk 5]
```diff
@@ -125,3 +130,68 @@ def __init__(
...
+
+        # Generate a dummy bbox
+        box = [48, 84, 73, 128]
+
+        if not framework == TensorType.PYTORCH:
+            raise NotImplementedError("Exporting LayoutLM to ONNX is currently only supported for PyTorch.")
+
+        if not is_torch_available():
+            raise ValueError("Cannot generate dummy inputs without PyTorch installed.")
+        import torch
```
Comment (NielsRogge) [inline src/transformers/models/layoutlm/configuration_layoutlm.py:188]:
```suggestion
        if not is_torch_available():
            raise ValueError("Cannot generate dummy inputs without PyTorch installed.")
            import torch
```
Shouldn't this have an indentation?

File: src/transformers/models/layoutlm/configuration_layoutlm.py [hunk 6]
```diff
@@ -125,3 +130,68 @@ def __init__(
...
+
+        input_dict["bbox"] = torch.tensor(
+            [
+                [0] * 4,
+                *[box] * seq_length,
+                [self.max_2d_positions] * 4,
+            ]
+        ).tile(batch_size, 1, 1)
+        return input_dict
```
Comment (NielsRogge) [inline src/transformers/models/layoutlm/configuration_layoutlm.py:197]:
```suggestion
        if not is_torch_available():
            raise ValueError("Cannot generate dummy inputs without PyTorch installed.")
            import torch

            input_dict["bbox"] = torch.tensor(
               [
                   [0] * 4,
                   *[box] * seq_length,
                   [self.max_2d_positions] * 4,
               ]
           ).tile(batch_size, 1, 1)
           return input_dict
```
Shouldn't this have an indentation?

**Comment by NielsRogge** [comment]
LGTM!

**Comment by LysandreJik** [comment]
This is making the LayoutLM ONNX test fail for the following reason:

```
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <transformers.models.layoutlm.configuration_layoutlm.LayoutLMOnnxConfig object at 0x7f344811ceb0>
tokenizer = PreTrainedTokenizerFast(name_or_path='microsoft/layoutlm-base-uncased', vocab_size=30522, model_max_len=512, is_fast=T...okens={'unk_token': '[UNK]', 'sep_token': '[SEP]', 'pad_token': '[PAD]', 'cls_token': '[CLS]', 'mask_token': '[MASK]'})
batch_size = -1, seq_length = -1, is_pair = False
framework = <TensorType.PYTORCH: 'pt'>

    def generate_dummy_inputs(
        self,
        tokenizer: PreTrainedTokenizer,
        batch_size: int = -1,
        seq_length: int = -1,
        is_pair: bool = False,
        framework: Optional[TensorType] = None,
    ) -> Mapping[str, Any]:
        """
        Generate inputs to provide to the ONNX exporter for the specific framework
    
        Args:
            tokenizer: The tokenizer associated with this model configuration
            batch_size: The batch size (int) to export the model for (-1 means dynamic axis)
            seq_length: The sequence length (int) to export the model for (-1 means dynamic axis)
            is_pair: Indicate if the input is a pair (sentence 1, sentence 2)
            framework: The framework (optional) the tokenizer will generate tensor for
    
        Returns:
            Mapping[str, Tensor] holding the kwargs to provide to the model's forward function
        """
    
        input_dict = super().generate_dummy_inputs(tokenizer, batch_size, seq_length, is_pair, framework)
    
        # Generate a dummy bbox
        box = [48, 84, 73, 128]
    
        if not framework == TensorType.PYTORCH:
            raise NotImplementedError("Exporting LayoutLM to ONNX is currently only supported for PyTorch.")
    
        if not is_torch_available():
            raise ValueError("Cannot generate dummy inputs without PyTorch installed.")
        import torch
    
>       input_dict["bbox"] = torch.tensor(
            [
                [0] * 4,
                *[box] * seq_length,
                [self.max_2d_positions] * 4,
            ]
        ).tile(batch_size, 1, 1)
E       RuntimeError: Trying to create tensor with negative dimension -1: [-1, 2, 4]

```

Could you take a look @nishprabhu @michaelbenayoun?

Will skip this test in the meantime.

**Comment by nishprabhu** [comment]
I think we have to import the compute_effective_axis_dimension function from the onnx module and compute the batch_size and seq_length dimensions in the configuration_layoutlm.py file. We are currently using -1 as the value for both dimensions which is causing the error. @michaelbenayoun @LysandreJik

**Pull request merged**